app/lib/kubernetes_client.py

```diff
@@ -6,20 +6,25 @@
 
 
 def list_stateful_sets():
-    return kubernetes_client.CustomObjectsApi().list_namespaced_custom_object(group="apps", version="v1", plural="statefulsets", namespace=namespace)['items']
+    return kubernetes_client.CustomObjectsApi().list_namespaced_custom_object(group="apps", version="v1",
+                                                                              plural="statefulsets",
+                                                                              namespace=namespace)['items']
 
 
 def list_validator_stateful_sets(role='authority'):
     stateful_sets = list_stateful_sets()
-    validator_stateful_sets = list(filter(lambda sts: sts['spec']['template']['metadata']['labels']['role'] == role, stateful_sets))
+    validator_stateful_sets = list(
+        filter(lambda sts: sts['spec']['template']['metadata']['labels'].get('role') == role, stateful_sets))
     return list(map(lambda sts: sts['metadata']['name'], validator_stateful_sets))
 
 
 def list_parachain_collator_stateful_sets(para_id):
     stateful_sets = kubernetes_client.CustomObjectsApi().list_namespaced_custom_object(group="apps", version="v1", plural="statefulsets", namespace=namespace)
-    collator_stateful_sets = list(filter(lambda sts: sts['spec']['template']['metadata']['labels']['role'] == 'collator' and
-                                                     sts['spec']['template']['metadata']['labels']['paraId'] == para_id,
-                                         stateful_sets['items']))
+
+    collator_stateful_sets = list(
+        filter(lambda sts: sts['spec']['template']['metadata']['labels'].get('role') == 'collator' and
+                           sts['spec']['template']['metadata']['labels'].get('paraId') == para_id,
```

Comment:
can we try to  use [get_parachain_id(pod)](https://github.com/paritytech/testnet-manager/blob/main/app/lib/parachain_manager.py#L77) it has paraId even if it is not set.

Reply:
After digging into it I think it can be very slow to have to query collators paraID from their RPC endpoints (1 request / collator). Especially as we do this in several places just to list the collators. As the manager doesn't have a database/cache this will impact the responsiveness. I think be rearchitecting some parts of the code and by adding caching it could be done.

Reply:
ok, it makes sense to leave it as is.